Conversions Page, Action Confirmation Modal Component

Deelane opened the pull request:
# Description

### Conversions Page

Conversions page built using the Units page from PR 787 as a template. Added delete functionality.

### Action Confirmation Modal Component

New modal component for confirming any action. This component can be used anywhere the developer needs confirmation before executing an action. Required props are a boolean "show" that controls the modal's visible state, a message to display in the modal, and 2 functions. One function is executed when the "confirm" button is pressed and one executes when the "reject" button is pressed. 

It is assumed that the developer implements their own functionality in the reject function for altering the "show" boolean to alter the visibility of the modal. Also, be sure to hide any parent modals too if you use the confirm modal as a nested modal. 

Optional props include confirm and reject text that will replace the default "yes" and "no" text for the buttons, and a title for the modal.

*Note: This modal component has very basic CSS, updates to the visuals to make it look nicer are welcome from any developer.

Fixes #700 to use identifier for meter.

## Type of change

- [ ] Note merging this changes the node modules
- [ ] Note merging this changes the database configuration.
- [x] This change requires a documentation update

## Checklist


- [x] I have followed the [OED pull request](https://openenergydashboard.github.io/developer/pr.html) ideas
- [x] I have removed text in ( ) from the issue request

abundyMB committed:
Partially completed conversion branch

abundyMB committed:
Merge branch 'abundy-conversion' into conversion

abundyMB committed:
Conversion update (conversion not stored in state)

abundyMB committed:
Conversion update (conversion not stored in state)

abundyMB committed:
Modified conversions w/ Delaney help (for testing)

abundyMB committed:
Conversions w/ basic working functionality

abundyMB committed:
Modified conversion edits to match using two IDs

abundyMB committed:
Conversion translation/added conversion router.get

abundyMB committed:
Conversion getById > getBySourceDestination

abundyMB committed:
Conversions successfully created

abundyMB committed:
Conversions for 7/27

abundyMB committed:
Conversion edit; only update modal on refresh

Deelane committed:
fixed edit unit not properly communicating with store

Deelane committed:
removed unused code

abundyMB committed:
Revert "fixed edit unit not properly communicating with store"

This reverts commit 38a2a6b6adf599fc5370a5a7f04d9e13ffeeb7ea.

abundyMB committed:
Conversion create/edit works as intended

abundyMB committed:
Verify new conversions don't already exist in DB

Deelane committed:
Added prototype conversion delete button, ConfirmActionModalComponent for confirming any action desired using a modal

Deelane committed:
Removed optional props, added component description

Deelane committed:
remove unused import

Deelane committed:
centered warning modal

Deelane committed:
fixed possibility of accessing units state before it loaded in conversions page

Deelane committed:
added bidirectional checks for valid conversions

Deelane committed:
comment fix

huss committed:
remove obsolete conversion route return item and route

huss committed:
Merge branch 'conversion' of https://github.com/Deelane/OED into conversion

Deelane committed:
fixed createconversion component possibly accessing undefined units state

Deelane committed:
Merge branch 'conversion' of https://github.com/Deelane/OED into conversion

huss committed:
Small conversion updates

- Nav button is Conversions (plural now).
- Help text and links unique for all conversion modals.
- Arrow in title is double ended if bidirectional.

huss committed:
Merge branch 'conversion' of https://github.com/Deelane/OED into conversion

Deelane committed:
Remove unused import

Deelane committed:
removed unused code

Deelane committed:
Fixed merge conflicts

Deelane committed:
git

Deelane committed:
fix comment typo

Deelane committed:
possible fix to possible accessing of units state before done loading

Deelane committed:
fix bug of create conversion possibly accessing nonexistent unit

huss committed:
TODO for no default unit on conversion create.

Deelane committed:
added "select a source/destination" as default for create conversion options, redundant check for negative ids in API validation

Deelane committed:
remove comment

Deelane committed:
hide default select option after selecting a source/destination unit

huss committed:
Merge branch 'resourceGeneralization' of https://github.com/OpenEnergyDashboard/OED into pr/Deelane/788

huss commented on src/client/app/components/ConfirmActionModalComponent.tsx:
```diff
@@ -0,0 +1,94 @@
+	actionFunction: () => void;
+}
+
+/**
```
This seems a general way that is only used in EditConversionModalComponent. Was it intended to be used elsewhere?

huss replied:
Added TODO for future consideration.

huss commented on src/server/routes/conversions.js:
```diff
@@ -0,0 +1,188 @@
+		res.status(400);
+	} else {
+		const conn = getConnection();
+		try {
```
Why do we need to get the conversion since all values overwritten? Is it to check if it exists?

huss replied:
Seems safe to assume conversion exists and the catch would return an error if fails. Changed to create a new conversion and update that without the first DB call.

huss commented on src/client/app/types/redux/conversions.ts:
```diff
@@ -0,0 +1,79 @@
+	hasBeenFetchedOnce: boolean,
+	isFetching: boolean;
+	selectedConversions: number[];
+	// Use an array of ConversionData because we need the combination of source/destination id to give us a unique conversion
```
I'm unclear on why this needs to be an array since each conversion has has both values. Can you help me understand?

huss replied:
I'm not sure if this needs to be an array but it works so letting it go for now so the PR can be cleared.

huss commented on src/client/app/utils/api/ConversionsApi.ts:
```diff
@@ -0,0 +1,37 @@
+	}
+
+	public async edit(conversion: ConversionData): Promise<ConversionEditData> {
+		return await this.backend.doPostRequest<ConversionEditData>(
```
The input is ConversionData but then it does the post with ConversionEditData.

huss replied:
Fixed when removed ConversionEditData.

huss commented on src/client/app/types/redux/conversions.ts:
```diff
@@ -0,0 +1,79 @@
+	note: string;
+}
+
+export interface ConversionEditData {
```
Edit seems the same as ConversionData. Is there a reason we need both? I know some others have both since the information varies (but not here). This is used in src/client/app/utils/api/ConversionsApi.ts for what is returned by edit (and posted) but SubmitEditedConversionAction uses ConversionData in the action above.

huss replied:
Removed ConversonEditData and replaced with ConversionData.

huss committed:
update PR

- fix up translations and standardize names; alphabetize.
- minor formatting & wording
- remove obsolete comment
- removed unused lodash import
- address couple of TODOs
- improve conversion route parameter checks

huss committed:
too many updates to PR

- Note some changes impacted other modals and were done to keep
everything consistent.
- Standardize CSS for modals by combining files. Note some of the items
defined may not be needed but waiting to here from others.
- Removed admin login checks from multiple pages since route does
not allow if not admin.
- Sorted units on create conversion dropdown menu.
- Changed how default value (-999) selected in create conversion
because it was causing console warnings.
- Added some translations that were needed.
- Reordered imports to try to group/standardize a little.
- Lots of formatting and minor edits.

huss committed:
display meters by identifier on dropdown menu

huss committed:
TODO on error found

See issue #790.

huss reviewed (changes requested):
This PR looks good. I have a few thoughts that I would like to see if the authors can give insight. I thank everyone for doing this major task.

huss commented:
I want to apologize to the authors of the PR for taking so long. I was hoping to clear up the issue of updating Cik on edits. However, it turned into a bigger issue and also impacts other pages. Thus, I have opened issue #791 to deal with this. This PR is passing local tests and is failing on GitHub per the ongoing issue. Once the last items are addressed and some final testing is done this should be ready to merge and then be put int development.

huss committed:
removed unused CSS styles

huss committed:
TODO to resolve PR comment

huss committed:
remove get of conversion to resolve PR comment

huss committed:
remove ConversionEditData for PR comment

huss commented:
All comments have been addressed. Note I updated the PR in several ways that are given in the commit messages. Local tests all pass so continuing to ignore GitHub Action failure (ongoing issue).

Note that there are issues in certain actions on conversions (and units/meters). Issue #791 was created to address these problems. However, this PR is now being merged to move it forward.
